### PR TITLE
OASIS-14: Make configprovider a component

### PR DESCRIPTION
**Pull request by mackjmr**

### CODE REVIEW

File: comp/otelcol/provider/go.mod [hunk 1]
```diff
@@ -0,0 +1,103 @@
+module provider
+
+go 1.21.6
```
Comment (mx-psi) [inline comp/otelcol/provider/go.mod:3]:
```suggestion
go 1.21.0
```

File: comp/otelcol/provider/providerimpl/go.mod [hunk 2]
```diff
@@ -1,6 +1,6 @@
-module provider
+module providerimpl
 
-go 1.21.0
+go 1.21.6
```
Comment (mx-psi) [inline comp/otelcol/provider/providerimpl/go.mod:3]:
```suggestion
go 1.21.0
```

File: go.mod [hunk 3]
```diff
@@ -1,6 +1,6 @@
 module github.com/DataDog/datadog-agent
 
-go 1.21.0
+go 1.21.6
```
Comment (mx-psi) [inline go.mod:3]:
```suggestion
go 1.21.0
```

**Review by mx-psi — COMMENTED** [review]
I don't have much context here, the changes look sensible to me but I guess someone from the squad should review this instead of me :)

File: comp/otelcol/provider/component.go [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package provider
+
+import "github.com/DataDog/datadog-agent/comp/otelcol/provider/providerimpl"
+
+// team: opentelemetry
+
+// Component is the component type.
+type Component interface {
+	providerimpl.ExtendedConfigProvider
```
Comment (dineshg13) [inline comp/otelcol/provider/component.go:9]: Can we define ExtendedConfigProvider in this file?

Can we follow new component structure ? use 
`provider/def` -> for component and other type definitions 
`provider/fx` -> to define module
`provider/impl` -> for implementation . 

Please see contribFx component or collector component

**Review by dineshg13 — CHANGES_REQUESTED** [review]

File: comp/otelcol/provider/component.go (hunk 4 above)
Comment (mackjmr) [reply comp/otelcol/provider/component.go]: Followed new component structure and Moved `ExtendedConfigProvider` definition to `provider/def`.

**Review by dineshg13 — APPROVED** [review]

File: comp/otelcol/provider/impl/provider.go [hunk 5]
```diff
@@ -3,13 +3,14 @@
...
 
-package provider
+package providerimpl
 
 import (
 	"context"
 	"fmt"
 	"os"
 
+	providerdef "github.com/DataDog/datadog-agent/comp/otelcol/provider/def"
```
Comment (dustmop) [inline comp/otelcol/provider/impl/provider.go:13]: Please remove the `def` from the aliased import:
```suggestion
	provider "github.com/DataDog/datadog-agent/comp/otelcol/provider/def"
```
We use only the component name (without "def") for the interface definition.

File: comp/otelcol/provider/fx/fx.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	providerimpl "github.com/DataDog/datadog-agent/comp/otelcol/provider/impl"
+	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
+
+	"go.uber.org/fx"
+)
+
+// Module defines the fx options for this component.
+func Module() fxutil.Module {
+	return fxutil.Component(
+		fx.Provide(providerimpl.NewConfigProvider))
```
Comment (dustmop) [inline comp/otelcol/provider/fx/fx.go:19]: Use the `ProvideComponentConstructor` helper function here:
```suggestion
		fxutil.ProvideComponentConstructor(providerimpl.NewConfigProvider))
```
This function guarantees non-fx constructors always work properly when used by fx. In this instance, it doesn't make a difference, but we prefer to always use this function in order to stay consistent across components using the new folder layout.

See this conversion guide for more info:
https://datadoghq.atlassian.net/wiki/spaces/ASC/pages/3737976969/Quick+Guide+to+Convert+to+multi-impl+Component#Create-fx-layer

File: cmd/otel-agent/subcommands/run/command.go [hunk 7]
```diff
@@ -30,7 +30,8 @@ import (
 	collectorfx "github.com/DataDog/datadog-agent/comp/otelcol/collector/fx"
 	"github.com/DataDog/datadog-agent/comp/otelcol/logsagentpipeline"
 	"github.com/DataDog/datadog-agent/comp/otelcol/logsagentpipeline/logsagentpipelineimpl"
-	configprovider "github.com/DataDog/datadog-agent/comp/otelcol/otlp/components/pipeline/provider"
+	providerdef "github.com/DataDog/datadog-agent/comp/otelcol/provider/def"
```
Comment (dustmop) [inline cmd/otel-agent/subcommands/run/command.go:33]:
```suggestion
	provider "github.com/DataDog/datadog-agent/comp/otelcol/provider/def"
```

File: cmd/otel-agent/subcommands/run/command.go [hunk 8]
```diff
@@ -85,9 +86,8 @@ func runOTelAgentCommand(_ context.Context, params *subcommands.GlobalParams, op
 		fetchonlyimpl.Module(),
 		collectorfx.Module(),
 		collectorcontribFx.Module(),
-		fx.Provide(configprovider.NewConfigProvider),
-		// For FX to provide the otelcol.ConfigProvider from the configprovider.ExtendedConfigProvider
-		fx.Provide(func(cp configprovider.ExtendedConfigProvider) otelcol.ConfigProvider {
+		providerfx.Module(),
+		fx.Provide(func(cp providerdef.ExtendedConfigProvider) otelcol.ConfigProvider {
```
Comment (dustmop) [inline cmd/otel-agent/subcommands/run/command.go:90]:
```suggestion
		fx.Provide(func(cp provider.ExtendedConfigProvider) otelcol.ConfigProvider {
```

File: comp/otelcol/provider/def/component.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+import (
+	"go.opentelemetry.io/collector/otelcol"
+)
+
+// team: opentelemetry
+
+// Component is the component type.
+type Component interface {
+	ExtendedConfigProvider
```
Comment (dustmop) [inline comp/otelcol/provider/def/component.go:16]: Why have two interfaces instead of just moving everything into this one?

Comment (mackjmr) [reply comp/otelcol/provider/def/component.go]: Moved everything into the Component interface 👍

File: comp/otelcol/provider/fx/fx.go (hunk 6 above)
Comment (mackjmr) [reply comp/otelcol/provider/fx/fx.go]: Switched to `fxutil.ProvideComponentConstructor`. This caught an issue where I wasn't using the `Requires` struct for my constructors argument, so I updated that as well.

**Comment by mackjmr** [comment]
Thanks for the thorough review @dustmop, latest commit should contain the request changes :)

**Review by dustmop — APPROVED** [review]

**Pull request merged**